File: pascal2.c
#include <stdio.h>
#define MAX 100
int cache[MAX][MAX];
int pascal(int row, int col) {
	if (cache[row][col] > 0) return cache[row][col];
	if (col == 0 || col == row) return cache[row][col] = 1;
	return cache[row][col] = pascal(row-1, col-1) + pascal(row-1,col);
}
int main(void) {
	for (int i = 0; i < MAX; i++)
		for (int j = 0; j < MAX; j++)
			cache[i][j] = 0;
	for (int i = 0; i < 10; i++) {
		for (int j = 0; j <= i; j++) {
			printf("%4d", pascal(i,j));
		}
		printf("\n");
	}
	return 0;
}

File: pascal2.js
function pascal(row, col) {
	if (col == 0 || row == col) return 1;
	var key = row + ',' + col;
	if (cache[key]) return cache[key];
	return cache[key] = pascal(row-1, col-1) + pascal(row-1, col);
}
var cache = [];
for (var i = 0; i < 10; i++) {
	var str = "";
	for (var j = 0; j <= i; j++)
		str += pascal(i, j) + " ";
	console.log(str);
}
